core/trino-main/src/test/java/io/trino/execution/scheduler/TestScaledWriterScheduler.java

```diff
@@ -0,0 +1,421 @@
+        }
+
+        public void setState(State state)
+        {
+            this.state = state;
+            if (stateChangeListener != null) {
+                stateChangeListener.stateChanged(state);
+            }
+        }
+
+        @Override
+        public State getState()
+        {
+            return state;
+        }
+
+        @Override
+        public void addStateChangeListener(StateMachine.StateChangeListener<State> stateChangeListener)
+        {
+            this.stateChangeListener = requireNonNull(stateChangeListener, "stateChangeListener is null");
+        }
+
+        @Override
+        public StageId getStageId()
+        {
+            return new StageId(new QueryId("id"), 0);
+        }
+
+        @Override
+        public int getAttemptId()
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void beginScheduling()
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void transitionToSchedulingSplits()
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public TaskLifecycleListener getTaskLifecycleListener()
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void schedulingComplete()
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void schedulingComplete(PlanNodeId partitionedSource)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void cancel()
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void abort()
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void recordGetSplitTime(long start)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Optional<RemoteTask> scheduleTask(InternalNode node, int partition, Multimap<PlanNodeId, Split> initialSplits)
+        {
+            return Optional.of(new TestingRemoteTask(TaskId.valueOf("taskId"), "nodeId", fragment));
+        }
+
+        @Override
+        public void failTask(TaskId taskId, Throwable failureCause)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public List<RemoteTask> getAllTasks()
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public List<TaskStatus> getTaskStatuses()
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Optional<ExecutionFailureInfo> getFailureCause()
+        {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    private static class TestingNodeSelector
```

Comment:
Why do we need this? Can't we directly use `UniformNodeSelector`?

Won't this work?

```
        private static final InternalNode NODE_1 = new InternalNode("node-1"..
        private static final InternalNode NODE_2 = new InternalNode("node-2"..

        new UniformNodeSelectorFactory(
            new InMemoryNodeManager(NODE_1, NODE_2),
            new NodeSchedulerConfig().setIncludeCoordinator(true),
            new NodeTaskMap(new FinalizerService())).createNodeSelector(testSessionBuilder().build(), Optional.empty())
```

Reply:
It doesn't work for the second `schedule` function call.

Reply:
Why is that?

Reply:
actually it works. thanks.